File: hw2/CalcTest.js
const Calc = require('./Calc');

// Simple test framework
function test(name, fn) {
    try {
        fn();
        console.log(`✓ ${name}`);
    } catch (error) {
        console.error(`✗ ${name}: ${error.message}`);
        process.exitCode = 1;
    }
}

function assertEqual(actual, expected, message) {
    if (actual !== expected) {
        throw new Error(`${message}: expected ${expected}, got ${actual}`);
    }
}

function assertThrows(fn, message) {
    try {
        fn();
        throw new Error(`${message}: expected to throw but didn't`);
    } catch (error) {
        if (error.message.includes("expected to throw")) {
            throw error;
        }
        // Expected exception thrown
    }
}

// Run tests
console.log("Running Calc tests...\n");

test("test_add", () => {
    assertEqual(Calc.add(2, 3), 5, "Addition failed");
});

// TDD Step 1: Test for subtract (FAILING TEST)
test("test_subtract", () => {
    assertEqual(Calc.subtract(5, 3), 2, "Subtraction failed");
});

// TDD Step 2: Test for multiply (FAILING TEST)
test("test_multiply", () => {
    assertEqual(Calc.multiply(4, 3), 12, "Multiplication failed");
});

// TDD Step 3: Test for divide (FAILING TEST)
test("test_divide", () => {
    assertEqual(Calc.divide(10, 2), 5.0, "Division failed");
});

// TDD Step 4: Test for divide by zero (edge case)
test("test_divide_by_zero", () => {
    assertThrows(() => Calc.divide(10, 0), "Divide by zero should throw error");
});

console.log("\nAll tests completed!");

File: hw2/CalcTest.java
// Introduction to Software Testing
// Authors: Paul Ammann & Jeff Offutt
// Chapter 3; page ??
// JUnit for Calc.java

import org.junit.*;
import static org.junit.Assert.*;

public class CalcTest
{
   @Test public void testAdd()
   {
      assertTrue ("Calc sum incorrect", 5 == Calc.add (2, 3));
   }
   
   // TDD Step 1: Test for subtract (FAILING TEST)
   @Test public void testSubtract()
   {
      assertTrue ("Calc subtract incorrect", 2 == Calc.subtract (5, 3));
   }
   
   // TDD Step 2: Test for multiply (FAILING TEST)
   @Test public void testMultiply()
   {
      assertTrue ("Calc multiply incorrect", 12 == Calc.multiply (4, 3));
   }
   
   // TDD Step 3: Test for divide (FAILING TEST)
   // Decision: divide returns double for precision
   @Test public void testDivide()
   {
      assertEquals ("Calc divide incorrect", 5.0, Calc.divide (10, 2), 0.001);
   }
   
   // TDD Step 4: Test for divide by zero (edge case)
   @Test(expected = IllegalArgumentException.class)
   public void testDivideByZero()
   {
      Calc.divide (10, 0);
   }
}
